includeAll not working with Liquibase 4.0.0

Starting with Liquibase 4.0.0, the includeAll definition in a change log file doesn´t work anymore if the referenced files are placed in a JAR file.
It´s working fine, if the files can be referenced from a target folder during development but not when the changesets should be accessed from JAR.

For details, see also https://forum.liquibase.org/t/includeall-with-v4-0-0/4655
Text from forum update
With the Quarkus version 1.7.1 which uses Liquibase 3.9.x everything works fine.

But Quarkus 1.8.x upgrades to Liquibase 4.0.0 and this causes issues with change sets using the includeAll tag.

Our changeLog.xml contains following:
<includeAll path="db/1.0.0/" errorIfMissingOrEmpty="true"/>

When we start the application in the Quarkus dev mode (mvn quarkus:dev) everything works fine; the changeLog is read and every file under the given path (db/1.0.0/) is executed/checked from Liquibase.

But if we run the application directly (with java -jar) we are getting following error:

```
Caused by: liquibase.exception.SetupException: Could not find directory or directory was empty for includeAll 'db/1.0.0/'
	at liquibase.changelog.DatabaseChangeLog.includeAll(DatabaseChangeLog.java:528)
```

The resources under this given folder are just not found.

If I define the includes per file, it´s working fine:

How can we define the includeAll tags with the new Liquibase 4.0.0 version?



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-714) by [Unito](https://www.unito.io/learn-more)
┆Fix Versions: Community 4.3.2,Liquibase 4.3.2

--- a/liquibase-core/src/main/java/liquibase/integration/spring/SpringResourceAccessor.java
+++ b/liquibase-core/src/main/java/liquibase/integration/spring/SpringResourceAccessor.java
@@ -98,7 +98,20 @@ public class SpringResourceAccessor extends AbstractResourceAccessor {
 
         //have to fall back to figuring out the path as best we can
         try {
-            return resource.getURL().toExternalForm().replaceFirst(".*!", "");
+            String url = resource.getURL().toExternalForm();
+            if (url.contains("!")) {
+                return url.replaceFirst(".*!", "");
+            } else {
+                while (!resourceLoader.getResource("classpath:"+url).exists()) {
+                    String newUrl = url.replaceFirst("^/?.*?/", "");
+                    if (newUrl.equals(url)) {
+                        throw new UnexpectedLiquibaseException("Could determine path for "+resource.getURL().toExternalForm());
+                    }
+                    url = newUrl;
+                }
+
+                return url;
+            }
         } catch (IOException e) {
             //the path gets stored in the databasechangelog table, so if it gets returned incorrectly it will cause future problems.
             //so throw a breaking error now rather than wait for bigger problems down the line
@@ -152,6 +165,7 @@ public class SpringResourceAccessor extends AbstractResourceAccessor {
     protected String finalizeSearchPath(String searchPath) {
         searchPath = "classpath:"+searchPath;
         searchPath = searchPath
+                .replaceAll("classpath\\*:", "classpath:")
                 .replace("\\", "/")
                 .replaceAll("//+", "/")
                 .replace("classpath:classpath:", "classpath:");

--- a/liquibase-core/src/main/java/liquibase/resource/ClassLoaderResourceAccessor.java
+++ b/liquibase-core/src/main/java/liquibase/resource/ClassLoaderResourceAccessor.java
@@ -121,11 +121,11 @@ public class ClassLoaderResourceAccessor extends AbstractResourceAccessor {
 
     protected String getFinalPath(String relativeTo, String streamPath) {
         streamPath = streamPath.replace("\\", "/");
-        streamPath = streamPath.replaceFirst("^classpath:", "");
+        streamPath = streamPath.replaceFirst("^classpath\\*?:", "");
 
         if (relativeTo != null) {
             relativeTo = relativeTo.replace("\\", "/");
-            relativeTo = relativeTo.replaceFirst("^classpath:", "");
+            relativeTo = relativeTo.replaceFirst("^classpath\\*?:", "");
             relativeTo = relativeTo.replaceAll("//+", "/");
 
             if (!relativeTo.endsWith("/")) {

